src/chart/helper/labelHelper.js

```diff
@@ -25,7 +25,8 @@ import {retrieveRawValue} from '../../data/helper/dataProvider';
  * @return {string} label string. Not null/undefined
  */
 export function getDefaultLabel(data, dataIndex) {
-    var labelDims = data.mapDimension('defaultedLabel', true);
+    var dataDimsLen = data.dimensions.length;
+    var labelDims = data.mapDimension(dataDimsLen > 2 ? data.dimensions[dataDimsLen - 1] : 'defaultedLabel', true);
```

Comment:
(1) the logic `dataDimsLen > 2 ? data.dimensions[dataDimsLen - 1] : ...` is to calculate a "default label" , so the code should be placed on where the `defaultedLabel` calculated rather here.
(2) We should not simply use the last dimension at the default label. Consider the dataset case:
```js
dataset: {
    source: [
       //  series0x series0y series1x series1y series2x series2y
           [11,              22,            33,       99,         33,          91          ],
           [121,              62,            93,       99,         83,          91          ],
           [16,              52,            33,       99,         33,          91          ],
    ] 
}
```
If using the last dimension as the label, all of the series will use the `series2y` as their label.

It a little difficulty to guess which dimension can be the default label.

Users can set label dimension by 
```js
encode: {label: 2}   means dimesion 2
```
or 
```js
formatter: '{@[2]}' // means dimesion 2
```

So do we need to fix that on code?
We might only fix #11528 on example?

Reply:
Or can we modify the bmap dimension #11939

Reply:
@100pah It's a change since we introduced `encode` in 4.0 . It breaks all `scatter` series. Not only the `bmap` one. 

In the 3.x version and earlier. We use the first two values to represent the coordinates and the third one as the value. It will be displayed in label and tooltip.

Reply:
@pissang 
> In the dataset case. encode is a necessary configuration.

`encode` is not necessary in `dataset` case.
```js
option = {
    xAxis: {},
    yAxis: {},
    dataset: {
        source: [
           //  series0x series0y series1x series1y series2x series2y
           [11,              22,            33,       99,         33,          91          ],
           [121,              62,            93,       99,         83,          91          ],
           [16,              52,            33,       99,         33,          91          ],
        ] 
    },
    series: [{
        type: 'scatter'
    }, {
        type: 'scatter'
    }, {
        type: 'scatter'
    }]
};
```
is a common usage.

A possible way for this issue could be: 
when we guessing which dimension can be the "default label", we detect that if the data is from dataset, use the current way, otherwise, if the data is from series itself (that means the data will not shared by other series), use the laster dimension as the label.

Reply:
> encode is not necessary in dataset case.

I mean encode is a very common used concept and configuration in the scene using dataset.  It's necessary when we want to use another dimension to represent the value besides the `x/y` dimension. 

But in the code without `dataset`, which is more common in our examples. It's not necessary to use encode to specify the value dimension in our earlier version. That's why our examples are broken.

> when we guessing which dimension can be the "default label", we detect that if the data is from dataset, use the current way, otherwise, if the data is from series itself (that means the data will not shared by other series), use the laster dimension as the label.

Perhaps it can solve the issue. Can you show a possible code to achieve this?

But I think the key issue is how to deal with the data from series and data from `dataset`. It seems currently we use the same encoding strategy on them. But they are different because data from `dataset` can be shared by different series.

Reply:
@pissang 
> So can we solve it by using the dimension that has not been encoded yet?

I am not understand what it means yet. Could you tell me more?

> It seems currently we use the same encoding strategy on them.

The attribute `encode` and the encode strategy is introduced to echarts earlier than `dataset`. It originally to enable users specify any columns to map x/y/labels/tooltips, rather than have to travel and exchange the order of the columns of the raw data. That especially helps the scenario that users need different types of mappings in a single case.

I think the concept and behavior of the attribute `encode` should better be consistent either using `series.data` or `dataset`. Otherwise it may bring confuse and more learning cost to users.

But for the "default strategy when no encode" I think probably we need to trade it differently for `series.data` and `dataset` if necessary.

@yufeng04 @pissang 
I think the fix could be:
(1) Add `isFromDataset` method to `List` class:
```js
+ listProto.isFromDataset = function () {
+     var provider = this._rawData;
+     // There might be no `getSource` in some legacy provider input from upper application.
+     return provider.getSource && provider.getSource().fromDataset;
+ };
```
(2) Modify the logic about "defaultedLabel" in `summarizeDimensions` of `dimensionHelper.js`:
```js

export function summarizeDimensions(data) {
    var summary = {};
    var encode = summary.encode = {};
    var notExtraCoordDimMap = createHashMap();
    var defaultedLabel = [];
    var defaultedTooltip = [];

    // See the comment of `List.js#userOutput`.
    var userOutput = summary.userOutput = {
        dimensionNames: data.dimensions.slice(),
        encode: {}
    };

+   var fromDataSet = this.isFromDataSet();
+   var defaultedLabelDimNonDataset;

    each(data.dimensions, function (dimName) {
        var dimItem = data.getDimensionInfo(dimName);

        var coordDim = dimItem.coordDim;
        if (coordDim) {
            if (__DEV__) {
                assert(OTHER_DIMENSIONS.get(coordDim) == null);
            }

            var coordDimIndex = dimItem.coordDimIndex;
            getOrCreateEncodeArr(encode, coordDim)[coordDimIndex] = dimName;

            if (!dimItem.isExtraCoord) {
                notExtraCoordDimMap.set(coordDim, 1);

                // Use the last coord dim (and label friendly) as default label,
                // because when dataset is used, it is hard to guess which dimension
                // can be value dimension. If both show x, y on label is not look good,
                // and conventionally y axis is focused more.
-               if (mayLabelDimType(dimItem.type)) {
+               if (fromDataset && mayLabelDimType(dimItem.type)) {
                    defaultedLabel[0] = dimName;
                }

                // User output encode do not contain generated coords.
                // And it only has index. User can use index to retrieve value from the raw item array.
                getOrCreateEncodeArr(userOutput.encode, coordDim)[coordDimIndex] = dimItem.index;
            }
            if (dimItem.defaultTooltip) {
                defaultedTooltip.push(dimName);
            }
        }

+       // For data specified on `series` directly, we use different strategy
+       // to guess defualt label from data specified on `dataset`. That is,
+       // Find the last possible dimension. For example, consider the case
+       // [[latitude, longitude, val], ...], the `val` would be the best choise.
+       if (!fromDataSet && !dimItem.isExtraCoord) {
+           defaultedLabelDimNonDataset = dimName;
+       }

        OTHER_DIMENSIONS.each(function (v, otherDim) {
            var encodeArr = getOrCreateEncodeArr(encode, otherDim);

            var dimIndex = dimItem.otherDims[otherDim];
            if (dimIndex != null && dimIndex !== false) {
                encodeArr[dimIndex] = dimItem.name;
            }
        });
    });

+   if (defaultedLabelDimNonDataset != null)
+       defaultedLabel[0] = defaultedLabelDimNonDataset;
+   }
```

Reply:
> So can we solve it by using the dimension that has not been encoded yet?

 For example, If we have three value in each item of `series.data`.
```js
[x, y, z]
```

Because the data is not shared by other series. After taking `[x, y]` and encoded to xAxis/yAxis. We can easily guess the third value `z` is an extra value which can be displayed as label and tooltip.  We can treat it as a function `z = f(x, y)`

But in the `dataset`, because each dimension may be taken by other series.  We can not simply use the third dimension,  or the last dimension. So can we take a further step and find if any dimension in the `dataset` has been taken by other series and being encoded?